Mathlib/Tactic/Congr!.lean

```diff
@@ -0,0 +1,248 @@
+-/
+def Lean.MVarId.userCongr? (mvarId : MVarId) : MetaM (Option (List MVarId)) :=
+  mvarId.withContext do
+    mvarId.checkNotAssigned `userCongr?
+    let some (lhs, _) := (← mvarId.getType').eqOrIff? | return none
+    let some name := lhs.getAppFn.cleanupAnnotations.constName? | return none
+    let congrTheorems := (← getSimpCongrTheorems).get name
+    -- Note: congruence theorems are in provided in decreasing order of priority.
+    for congrTheorem in congrTheorems do
+      let res ← observing? do
+        let cinfo ← getConstInfo congrTheorem.theoremName
+        let us ← cinfo.levelParams.mapM fun _ => mkFreshLevelMVar
+        let proof := mkConst congrTheorem.theoremName us
+        let ptype ← instantiateTypeLevelParams cinfo us
+        applyCongrThm? mvarId ptype proof
+      if let some mvars := res then
+        return mvars
+    return none
+
+/--
+Try to convert an `Iff` into an `Eq` by applying `iff_of_eq`.
+If successful, returns the new goal, and otherwise returns the original `MVarId`.
+-/
+def Lean.MVarId.iffOfEq (mvarId : MVarId) : MetaM MVarId := do
+  let res ← observing? do
+    let [mvarId] ← mvarId.apply (mkConst `iff_of_eq []) | failure
+    return mvarId
+  return res.getD mvarId
+
+/--
+Try to convert an `Eq` into an `Iff` by applying `propext`.
+If successful, then returns then new goal, otherwise returns the original `MVarId`.
+-/
+def Lean.MVarId.propext (mvarId : MVarId) : MetaM MVarId := do
+  let res ← observing? do
+    let [mvarId] ← mvarId.apply (mkConst `propext []) | failure
+    return mvarId
+  return res.getD mvarId
+
+/-- Helper theorem for `LEan.MVar.liftReflToEq`. -/
+theorem Lean.MVarId.rel_of_eq_and_refl {R : α → α → Prop} (hxy : x = y) (h : R x x) :
+    R x y := hxy ▸ h
+
+/--
+Use a `refl`-tagged lemma to convert the goal into an `Eq`. If this can't be done, returns
+the original `MVarId`.
+-/
+def Lean.MVarId.liftReflToEq (mvarId : MVarId) : MetaM MVarId := do
+  mvarId.checkNotAssigned `liftReflToEq
+  let tgt ← withReducible <| mvarId.getType'
+  let .app (.app rel _) _ := tgt | return mvarId
+  if rel.isAppOf `Eq then
+    -- No need to lift Eq to Eq
+    return mvarId
+  let reflLemmas ← (Mathlib.Tactic.reflExt.getState (← getEnv)).getMatch rel
+  for lem in reflLemmas do
+    let res ← observing? do
+      -- First create an equality relating the LHS and RHS
+      -- and reduce the goal to proving that LHS is related to LHS.
+      let [mvarIdEq, mvarIdR] ←
+            mvarId.apply (← mkConstWithFreshMVarLevels ``Lean.MVarId.rel_of_eq_and_refl)
+        | failure
+      -- Then fill in the proof of the latter by reflexivity.
+      let [] ← mvarIdR.apply (← mkConstWithFreshMVarLevels lem) | failure
+      return mvarIdEq
+    if let some mvarId := res then
+      return mvarId
+  return mvarId
+
+/--
+Try to close the goal using `Subsingleton.elim`. Returns whether or not it succeeds.
+-/
+def Lean.MVarId.subsingletonElim (mvarId : MVarId) : MetaM Bool := do
```

Comment:
Can we do this without assigning universe metavariables? That regression seems very unfortunate.

Reply:
I think I found a straightforward way to do it. It synthesizes a `Subsingleton.elim` term using `mkAppM`, which uses `withNewMCtxDepth` so outside metavariables can't be reassigned.